contracts/JBSwapTerminal.sol

```diff
@@ -0,0 +1,440 @@
+            _minimumReceivedFromSwap = _getTwapFrom(_pool, _amount);
+        }
+
+        // Swap (will check if we're withing the slippage tolerance in the callback))
+        uint256 _receivedFromSwap = _swap(_pool, _amount, _minimumReceivedFromSwap);
+
+        // Unwrap weth if needed
+
+        // Pay on primary terminal, with correct beneficiary (sender or benficiary if passed)
+        _terminal.pay{value: _token == JBToken.ETH ? _receivedFromSwap : 0}(
+            _projectId,
+            _token,
+            _receivedFromSwap,
+            _beneficiary,
+            _minReturnedTokens,
+            _memo,
+            _metadata
+        );
+
+    }
+
+    /// @notice 
+    /// @param _projectId The ID of the project to which the funds received belong.
+    /// @param _amount The amount of tokens to add, as a fixed point number with the same number of decimals as this terminal. If this is an ETH terminal, this is ignored and msg.value is used instead.
+    /// @param _token The token being paid. This terminal ignores this property since it only manages one currency.
+    /// @param _shouldRefundHeldFees A flag indicating if held fees should be refunded based on the amount being added.
+    /// @param _memo A memo to pass along to the emitted event.
+    /// @param _metadata Extra data to pass along to the emitted event.
+    function addToBalanceOf(
+        uint256 _projectId,
+        address _token,
+        uint256 _amount,
+        bool _shouldRefundHeldFees,
+        string calldata _memo,
+        bytes calldata _metadata
+    ) external payable virtual override {
+        // Check the terminal accepts the token
+
+        // Accept the funds.
+        _amount = _acceptFundsFor(_projectId, _token, _amount, _metadata);
+
+        // Check for quote
+
+        // If no quote, check there is a pool assigned and get a twap
+
+        // Try to swap, if fails, revert
+
+        // Unwrap weth if needed
+
+        // Add to balance on primary terminal
+    }
+
+    /// @notice 
+    /// @param _holder The account to redeem tokens for.
+    /// @param _projectId The ID of the project to which the tokens being redeemed belong.
+    /// @param _tokenCount The number of project tokens to redeem, as a fixed point number with 18 decimals.
+    /// @param _token The token being reclaimed. This terminal ignores this property since it only manages one token.
+    /// @param _minReturnedTokens The minimum amount of terminal tokens expected in return, as a fixed point number with the same amount of decimals as the terminal.
+    /// @param _beneficiary The address to send the terminal tokens to.
+    /// @param _metadata Bytes to send along to the data source, delegate, and emitted event, if provided.
+    /// @return reclaimAmount The amount of terminal tokens that the project tokens were redeemed for, as a fixed point number with 18 decimals.
+    function redeemTokensOf(
```

Comment:
out of scope. 
lets scope this specifically to `pay`, and `addToBalance` if we're feeling generous.
The main value prop here is to facilitate fee collection for any number of coins. 

let's widen scope to redeem in december if we get to it, alongside maybe sell-back-delegate.

Reply:
meh, even redeem-side-swap, and sell-back-delegate seem unnecessary overall. not convinced they'll ever really be useful